atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/ShardedSweepableBucketRetriever.java

```diff
@@ -0,0 +1,116 @@
+    private final SweeperStrategy strategy; // TODO: Maybe this should be a parameter on getSweepableBuckets instead.
+    // It depends on whether we want a _single_ coordinator to handle both thorough and conservative sweep, or separate
+
+    private final ShardedRetrievalStrategy shardedRetrievalStrategy;
+    private final ShardedSweepTimestampManager sweepTimestampManager;
+    private final ParallelTaskExecutor parallelTaskExecutor;
+
+    private final Refreshable<Integer> maxParallelism;
+
+    private final Refreshable<Duration> minimumBackoff;
+    private final Refreshable<Duration> maxJitter;
+
+    private ShardedSweepableBucketRetriever(
+            Refreshable<Integer> numShards,
+            SweeperStrategy strategy,
+            ShardedRetrievalStrategy shardedRetrievalStrategy,
+            ShardedSweepTimestampManager sweepTimestampManager,
+            ParallelTaskExecutor parallelTaskExecutor,
+            Refreshable<Integer> maxParallelism,
+            Refreshable<Duration> minimumBackoff,
+            Refreshable<Duration> maxJitter) {
+        this.numShards = numShards;
+        this.shardedRetrievalStrategy = shardedRetrievalStrategy;
+        this.strategy = strategy;
+        this.sweepTimestampManager = sweepTimestampManager;
+        this.parallelTaskExecutor = parallelTaskExecutor;
+        this.maxParallelism = maxParallelism;
+        this.minimumBackoff = minimumBackoff;
+        this.maxJitter = maxJitter;
+    }
+
+    public static SweepableBucketRetriever create(
+            Refreshable<Integer> numShards,
+            SweeperStrategy strategy,
+            ShardedRetrievalStrategy shardedRetrievalStrategy,
+            ShardedSweepTimestampManager sweepTimestampManager,
+            ParallelTaskExecutor parallelTaskExecutor,
+            Refreshable<Integer> maxParallelism,
+            Refreshable<Duration> minimumBackoff,
+            Refreshable<Duration> maxJitter) {
+        return new ShardedSweepableBucketRetriever(
+                numShards,
+                strategy,
+                shardedRetrievalStrategy,
+                sweepTimestampManager,
+                parallelTaskExecutor,
+                maxParallelism,
+                minimumBackoff,
+                maxJitter);
+    }
+
+    @Override
+    public Set<SweepableBucket> getSweepableBuckets() {
+        List<List<SweepableBucket>> sweepableBuckets;
+        try (CloseableTracer tracer = CloseableTracer.startSpan("getSweepableBucketsAcrossAllShards")) {
+            // TODO: Time it!
+            sweepableBuckets = parallelTaskExecutor.execute(
+                    IntStream.range(0, numShards.get()).boxed(),
+                    this::getSweepableBucketsForShardWithJitter,
+                    maxParallelism.get());
+        }
+        return sweepableBuckets.stream().flatMap(List::stream).collect(Collectors.toSet());
+    }
+
+    // Allows us to slow down / speed up the rate of requests to the underlying storage.
```

Comment:
Hmm. This makes the sleep part of the task - is it intentional that we hold the semaphores when performing this backoff?